libcst/matchers/_matcher_base.py

```diff
@@ -344,6 +357,46 @@ def _match_func(value: object) -> bool:
     return MatchIfTrue(_match_func)
 
 
+class MatchMetadata:
+    """
+    Matcher that matches if for each metadata provider in the mapping, the node's
+    metadata matches the mapping value.
+
+    TODO: Docs.
+
+    This can be used in place of any concrete matcher as long as it is not the
+    root matcher. Calling :func:`matches` directly on a :class:`MatchMetadata` is
+    redundant since you can just check the metadata on the root node that you
+    are passing to :func:`matches`.
+    """
+
+    # pyre-ignore There is no Mapping in python types that is covariant in the
+    # key, which is what we need here. So, lets loosen the type.
+    def __init__(self, metadata: Mapping[Type[Any], object]) -> None:
+        self.metadata: Mapping[meta.ProviderT, object] = metadata
+
+    def __or__(self, other: _OtherNodeT) -> "OneOf[Union[MatchMetadata, _OtherNodeT]]":
+        # Without the cast, pyre thinks that this returns OneOf[object]
+        return cast(OneOf[Union[MatchMetadata, _OtherNodeT]], OneOf(self, other))
+
+    def __and__(self, other: _OtherNodeT) -> "AllOf[Union[MatchMetadata, _OtherNodeT]]":
+        # Without the cast, pyre thinks that this returns OneOf[object]
```

Comment:
I presume this is a copy-paste foul and should read `AllOf[object]`

Reply:
Yerp, will fix when I iterate on this.